src/com/gfredericks/schpec.clj

```diff
@@ -32,3 +32,15 @@
            (limit-keys ks# ~all-keys)
            ks#)
          (fn [] (sg/fmap (fn [m#] (select-keys m# ~all-keys)) (s/gen ks#)))))))
+
+(defmacro xor
+  "Like [[s/or]], but values can only match exactly one spec."
+  [& key-pred-forms]
+  (let [key-pred-forms (set (partition 2 key-pred-forms))
+        opts (mapcat
+              (fn [[name spec :as pair]]
+                (let [preds (for [s (map second (disj key-pred-forms pair))]
+                              `(fn [x#] (not (s/valid? ~s x#))))]
+                  [name `(s/and ~spec ~@preds)]))
+              key-pred-forms)]
+    (eval `(s/or ~@opts))))
```

Comment:
More specifically, ISTR that in the other place what I was trying to accomplish (and failed :/) was to get spec's explain to produce better error messages when something isn't in a particular set, but because it's a macro, the fn literal would just show the name of the set rather than its contents (which is of course much more interesting).